### PR TITLE
Implementation of complex Data formats

### PR DESCRIPTION
**Pull request by alexanderst**
Implementing construction and ser-de for :
- NDArray
- Image
- metadata (Data inside Data objects)

### CODE REVIEW

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 1]
```diff
@@ -93,8 +106,130 @@ else if (value.type().equals(ValueType.NDARRAY)) {
...
+                        build();
+            }
+            else if (value.type() == ValueType.BOUNDING_BOX) {
+                BoundingBox boundingBox = (BoundingBox) nextItem.getValue().get();
+                DataProtoMessage.BoundingBox pbBox = null;
+                if (boundingBox instanceof BBoxCHW) {
+                    pbBox = DataProtoMessage.BoundingBox.newBuilder().
+                            setCx(boundingBox.cx()).setCy(boundingBox.cy()).
+                            setH(boundingBox.y2()-boundingBox.y1()).
+                            setW(boundingBox.x2()-boundingBox.x1()).
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:128]: I've merged PR today with height() and width() methods on BoundingBox interface

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 2]
```diff
@@ -93,8 +106,130 @@ else if (value.type().equals(ValueType.NDARRAY)) {
...
+                        pbArrays.add(pbNDArray);
+                    }
+                    DataProtoMessage.NDArrayList toAdd = DataProtoMessage.NDArrayList.newBuilder().addAllList(pbArrays).build();
+                    DataProtoMessage.List toAddGen = DataProtoMessage.List.newBuilder().setNdList(toAdd).build();
+                    item = DataProtoMessage.DataScheme.newBuilder().
+                            setListValue(toAddGen).
+                            setListTypeValue(ValueType.NDARRAY.ordinal()).
+                            setTypeValue(ValueType.LIST.ordinal()).
+                            build();
+                }
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:231]: Let's do i`f(...){ } else if(...){ } else { throw new RuntimeException("List type not yet supported/implemented: " + ...)}`
At least if we add new dtype and forget to update this method we get to know immediately rather than get our data going missing silently (which could be quite difficult to debug in say a distributed environment)


Also (separate issue) - no `List<BoundingBox>` yet.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 3]
```diff
@@ -126,14 +261,74 @@ public static Data convertProtobufToData(DataProtoMessage.DataMap dataMap) {
...
+                Data embeddedData = deserialize(itemMetaData);
+                retData.put(entry.getKey(), embeddedData);
+            }
+            if (item.getTypeValue() == DataProtoMessage.DataScheme.ValueType.BOUNDING_BOX.ordinal()) {
+                DataProtoMessage.BoundingBox pbBox = item.getBoxValue();
+                BoundingBox boundingBox = null;
+                if (pbBox.getType() == DataProtoMessage.BoundingBox.BoxType.CHW)
+                    boundingBox = new BBoxCHW(pbBox.getCx(), pbBox.getCy(), pbBox.getH(), pbBox.getW());
+                else if (pbBox.getType() == DataProtoMessage.BoundingBox.BoxType.XY)
+                    boundingBox = new BBoxXY(pbBox.getX0(), pbBox.getX1(), pbBox.getY0(), pbBox.getY1());
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:275]: Missing the probabability and label fields here.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 4]
```diff
@@ -126,14 +261,74 @@ public static Data convertProtobufToData(DataProtoMessage.DataMap dataMap) {
...
+                }
             }
             if (item.getTypeValue() == DataProtoMessage.DataScheme.ValueType.IMAGE.ordinal()) {
                 DataProtoMessage.Image pbImage = item.getImValue();
                 List<ByteString> pbData = pbImage.getDataList();
                 byte[] data = pbData.get(0).toByteArray();
-                retData.put(entry.getKey(), Image.create(data));
+                // TODO: obviously should be working for different formats
+                Png png = new Png(data);
+                retData.put(entry.getKey(), Image.create(png));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:331]: Let's add a `string type` field to protobuf, and then check it's "PNG" here (exception if it's anything else for now)
```
message Image {
    string type = 1;
    repeated bytes data = 2;
}
```

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 5]
```diff
@@ -149,6 +344,7 @@ public static Data convertProtobufToData(DataProtoMessage.DataMap dataMap) {
                 ByteBuffer bb = ByteBuffer.wrap(bytes);
                 // TODO: fix ndarray type
                 SerializedNDArray ndArray = new SerializedNDArray(NDArrayType.FLOAT, aShapes, bb);
+                retData.put(entry.getKey(), NDArray.create(ndArray));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:347]: FYI types are all here: https://github.com/KonduitAI/konduit-serving/blob/master/konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/data/NDArrayType.java

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 4 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:394]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 3 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 1 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 2 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:374]: Fixed both.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 5 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 6]
```diff
@@ -326,17 +327,20 @@ public void putList(String key, List<?> data, ValueType vt){
...
+        DataValue metaData = (DataValue) dataMap.get(Data.RESERVED_KEY_METADATA);
+        if (metaData == null)
+            return null;
+        return metaData.get();
     }
 
     @Override
     public void setMetaData(Data data) {
-        this.metaData = data;
+        dataMap.put(Data.RESERVED_KEY_METADATA, new DataValue(data));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java:343]: Please revert these changes
Storing in protobuf like this is totally fine, but not at Java level. Let's keep the metaData field.
To do it this way will required filtering the keys to remove the metadata key for methods like keys(), etc.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 7]
```diff
@@ -18,22 +18,175 @@
...
 import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
 
 public class ProtobufUtils {
 
-    public static DataProtoMessage.DataMap convertJavaToProtobufData(Map<String,Value> dataMap) {
+    private static String PNG = "PNG";
+
+    private static List<ByteString> ndArrayToByteStringList(SerializedNDArray sn) {
+        byte[] bufferBytes = new byte[sn.getBuffer().remaining()];
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:43]: Let's add a `sn.getBuffer().rewind();` before we use it here to make sure we get all values.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java (hunk 6 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Appears to be not so small in protobuf ser-de. Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java (hunk 7 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:43]: Fixed.
